[ML] Regression dependent variable must be numeric

dimitris-athanasiou opened the pull request:
This adds a validation that the dependent variable of a regression
analysis must be numeric.

dimitris-athanasiou committed:
[ML] Regression dependent variable must be numeric

This adds a validation that the dependent variable of a regression
analysis must be numeric.

przemekwitek commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/Regression.java:
```diff
@@ -201,8 +202,8 @@ public boolean supportsCategoricalFields() {
-    public Set<String> getRequiredFields() {
-        return Collections.singleton(dependentVariable);
+    public List<RequiredField> getRequiredFields() {
+        return Collections.singletonList(new RequiredField(dependentVariable, new TreeSet<>(Types.numerical())));
```
Since you are already using `Collections.singletonList` in this line, could you also use `Collections.singleton` instead of `new TreeSet<>`?

dimitris-athanasiou replied:
Actually, I can simply pass `Types.numerical()` directly. I'll do that.

przemekwitek commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/RequiredField.java:
```diff
@@ -0,0 +1,31 @@
+public class RequiredField {
+
+    private final String name;
+    private final SortedSet<String> types;
```
Why does it need to be sorted? Please add a comment.

dimitris-athanasiou replied:
It is to ensure types are sorted in error messages. I'll add a comment.

przemekwitek reviewed (approved):
LGTM

dimitris-athanasiou committed:
Address review comments and fix some problems

In addition to addressing the review comments, this
commit fixes a few issues I found during testing.

In particular:

- if there were mappings for required fields but they were
not included we were not reporting the error
- if explicitly included fields had unsupported types we were
not reporting the error

Unfortunately, I couldn't get those fixed without refactoring
the code in `ExtractedFieldsDetector`.

dimitris-athanasiou commented:
@przemekwitek I have pushed a commit that addresses your feedback. However, the commit also fixes a couple other issues and it needs its own review. Would you please have a look?

przemekwitek commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/extractor/ExtractedFieldsDetector.java:
```diff
@@ -202,16 +191,64 @@ private void includeAndExcludeFields(Set<String> fields) {
+
+            // Check all field has compatible type
+            Set<String> asSet = new HashSet<>(Collections.singleton(field));
+            removeFieldsWithIncompatibleTypes(asSet);
```
Personally, I would extract out the part of `removeFieldsWithIncompatibleTypes` that checks the types and re-use the extracted method here and in `removeFieldsWithIncompatibleTypes`. It would allow for more direct statement, like:
```
if (!fieldHasCompatibleTypes(field)) {
  throw ...
}
```

dimitris-athanasiou replied:
Will do

przemekwitek commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/extractor/ExtractedFieldsDetector.java:
```diff
@@ -80,33 +69,33 @@
+        checkNoIgnoredFields(fields);
+        checkFieldsHaveCompatibleTypes(fields);
+        checkRequiredFields(fields);
+        checkResultsFieldIsNotPresent();
```
It performs the check whether or not `resultsField` is `null`. Is that ok? Do we have a test for that?

dimitris-athanasiou replied:
Yes, it should be ok. We just don't want to perform this validation when the resultsField is null. There is a test: `testDetect_NullResultsField`. You've added it ;-)

dimitris-athanasiou replied:
I see your point. I have removed the call to `checkResultsFieldIsNotPresent` from `detect()` because it is redundant.

przemekwitek commented on x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/dataframe/extractor/ExtractedFieldsDetectorTests.java:
```diff
@@ -146,7 +146,55 @@ public void testDetect_GivenRegressionAndRequiredFieldMissing() {
+            .addAggregatableField("some_keyword", "keyword")
+            .addAggregatableField("foo", "float")
+            .build();
+        FetchSourceContext analyzedFields = new FetchSourceContext(true, new String[]  {"some_float", "some_keyword"}, new String[0]);
```
What if `foo` was in both `included` and `excluded`?

dimitris-athanasiou replied:
`excluded` prevails. I'll add a test.

przemekwitek reviewed (approved):
@dimitris-athanasiou LGTM after you change although `ExtractedFieldsDetector` class is a bit hard to follow. I don't have an immediate remedy to that though :/

dimitris-athanasiou committed:
More review feedback

dimitris-athanasiou committed:
Remove redundant call of `checkResultsFieldIsNotPresent`

dimitris-athanasiou commented:
run elasticsearch-ci/1
